(#456) Add io_apis_test IO exerciser. Minor cleanup of laio.c

gapisback opened the pull request:
This commit adds a new `io_apis_test` target under `functional/driver_test`.

This new test exercises sync IO APIs and async reads APIs, as a way to do minimal validation of the IO sub-system and its APIs. 

Minor code & comment cleanup is done in `laio.c` to improve readability and understanding of this module. 

Few existing error & assertion messages reporting IO errors are enhanced to report more diagnostic data that lead to said errors. Most of these are around I/O calls and seen while opening a SplinterDB device.

**No new functional changes** are introduced with this commit.

Most of the changes in `laio.c` add clarifying comments / fn-prologs to improve overall readability. 
Some existing asserts are expanded to print additional diagnostic data to improve troubleshooting of IO errors (which are, by nature, difficult to diagnose).

gapisback commented on src/clockcache.c:
```diff
@@ -3054,14 +3054,46 @@ clockcache_print(platform_log_handle *log_handle, clockcache *cc)
+      retval =
+         ((base_addr != 0) && (allocator_get_ref(cc->al, base_addr) != 0));
+      if_fails = TRUE;
+      line = __LINE__;
```
This function can return `FALSE` from either side of the `if()`.

These diagnostics tease apart the piece that is failing, and print appropriate diagnostics data to simplify troubleshooting which side of these conditions failed.

rtjohnso replied:
I find this very confusing.  Here's a simpler approach:

1. Restore the original control flow.
2. Move the `base_addr != 0` into the `if` condition (`if (base_addr != 0 && addr < allocator_get_capacity(cc->al)) {`.
3. In the true branch of the `if`, store the refcount in a local variable.  If it is 0, print a diagnostic. Then `return refcount != 0;`
4. In the false branch of the `if`, print a diagnostic of all the relevant variables and return FALSE.

gapisback replied:
@rtjohnso - Reworked per your suggestion. It's much cleaner now. thanks!

gapisback commented on src/io.h:
```diff
@@ -75,11 +77,13 @@ typedef struct io_ops {
    io_cleanup_all_fn         cleanup_all;
    io_thread_register_fn     thread_register;
    io_max_latency_elapsed_fn max_latency_elapsed;
+   io_get_context_fn         get_context;
```
Added helper `get` method to return IO-Context. 

This is used in tests. (Currently, it's the same for all threads, but in some new development, this IO-context could be thread-specific. In that scenario, this kind of `get` method helps clarify the testing logic.)

gapisback commented on src/io.h:
```diff
@@ -75,11 +77,13 @@ typedef struct io_ops {
 // to sub-class io, make an io your first field;
 struct io_handle {
    const io_ops *ops;
+   int           nbytes_rw; // # of bytes read / written by IO call.
```
Added for diagnostics. The IO call returns the # of bytes read/written.

Diagnostic messages to trace incomplete IO will report this value, as a troubleshooting aid.

gapisback replied:
@ajhconway - I ran into some IO errors while doing other integration work and test-dev, so chased this down to understand why the IO failed.

The new field is supposed to return the # of bytes read / written on an IO call that failed.

Renamed it as you suggested.

About raising an error message: All places where `io_read()` or `io_write()` is done currently do a `platform_assert()`. I did not try to change them all to report an error message in this rework. I feel that is something that could be parceled out (if deemed important) into a follow-on issue.

gapisback commented on src/platform_linux/laio.c:
```diff
@@ -4,7 +4,14 @@
+ *
+ * - Sync  IO interfaces: io_read(), io_write()
+ * - Async IO interfaces: io_read_async(), io_write_async()
+ * - Async IO completion interfaces: io_cleanup(), io_cleanup_all()
```
I think this comment block doc'ing sync / async interfaces is reasonably complete.

If you'd like to see more doc'ed, kindly list the things that should be included.

gapisback replied:
Added these comments to help better doc the API.

gapisback commented on src/platform_linux/laio.c:
```diff
@@ -91,19 +113,40 @@ io_handle_init(laio_handle         *io,
+                   io->fd,
+                   strerror(errno));
+
+   if (is_create) {
```
This fixes a latent bug that will trip and makes the assert on L123 moot.

If `open()` on L19 were to fail, `fallocate()` on old L96 will also seg-fault.

Rearranged code to make it a bit more fail-safe.

gapisback commented on src/platform_linux/laio.c:
```diff
@@ -164,10 +221,17 @@ laio_write(io_handle *ioh, void *buf, uint64 bytes, uint64 addr)
+ * Return a ptr to the k'th Async IO request structure, accounting
+ * for a nested array of 'async_max_pages' pages of IO vector structures
+ * at the end of each Async IO request structure.
+ */
```
Clarified workings thru this prolog.

gapisback commented on src/platform_linux/laio.c:
```diff
@@ -292,21 +380,35 @@ laio_cleanup(io_handle *ioh, uint64 count)
    io = (laio_handle *)ioh;
+
+   // Check for completion of up to 'count' events, one event at a time.
+   // Or, check for all outstanding events (count == 0)
```
No code changes in this busy diff block. Added comments, and elaborated data printed when assert fails, below.

gapisback commented on tests/functional/io_apis_test.c:
```diff
@@ -0,0 +1,831 @@
+
+/*
+ * ----------------------------------------------------------------------------
+ * splinter_io_apis_test.c --
```
New test case to exercise core IO APIs.

For now, this is not adding any great testing value, other than added value to my understanding of the IO sub-system.

If ever we bring-on a new platform [ E.g. Async-IO on Mac/OSX] this test will provide regression cover.

gapisback commented on test.sh:
```diff
@@ -527,6 +527,8 @@ function run_fast_unit_tests() {
    "$BINDIR"/unit/limitations_test
    "$BINDIR"/unit/task_system_test
+
+   "$BINDIR"/driver_test io_apis_test
```
New test is exercised here.

ajhconway reviewed (changes requested):
I like most of this.

I think the `nbytes_rw` should go away in favor of more message printing, unless there's a reason that won't work.

rtjohnso commented on src/platform_linux/laio.c:
```diff
@@ -19,36 +26,50 @@
 laio_read(io_handle *ioh, void *buf, uint64 bytes, uint64 addr);
-platform_status
+
+static platform_status
```
If, outside this file, they are referenced only through function pointers, then they should be `static`.

`static` controls the visibility only of the _name_.  But you can still call the function through other means (i.e. function pointers).

rtjohnso replied:
And I totally support making things static whenever possible, so, thank you!

gapisback replied:
I verified that various builds still succeed after making these `static` . Verified on Docker-Mac with `gcc` and `clang` also. I think this is a preferable change to get in.

rtjohnso commented on tests/functional/io_apis_test.c:
```diff
@@ -0,0 +1,831 @@
+ * functions invoked by pthreads to invoke the work-horse function(s).
+ */
+typedef struct io_test_fn_args {
+   platform_heap_id    arg_hid;
```
We don't prefix fields in the rest of the code.  Please follow the general naming conventions.

gapisback replied:
Renamed member fields as suggested.

rtjohnso commented on tests/functional/io_apis_test.c:
```diff
@@ -0,0 +1,831 @@
+   platform_heap_id    arg_hid;
+   io_config          *arg_io_cfgp;
+   platform_io_handle *arg_io_hdlp;
+   io_context_t        arg_io_ctxt_t; // Expected opaque context handle
```
The `_t` suffix indicates that a name refers to a type.  Do not put it on things of that type.

gapisback replied:
Done; removed. Renamed to `io_ctxt`.

rtjohnso commented on tests/functional/io_apis_test.c:
```diff
@@ -0,0 +1,831 @@
+typedef void (*test_io_thread_hdlr)(void *arg);
+
+/* SplinterDB device size; small one is good enough for IO APIs testing */
+#define SPLINTER_DEVICE_SIZE_MB 128
```
Maybe just `DEVICE_SIZE_MB` since there's nothing about splinter in here?

gapisback replied:
Done.

gapisback commented on src/clockcache.c:
```diff
@@ -3054,14 +3054,46 @@ clockcache_print(platform_log_handle *log_handle, clockcache *cc)
 {
-   if (addr % clockcache_page_size(cc) != 0)
+   if (addr % clockcache_page_size(cc) != 0) {
+      platform_error_log("%s():%d: addr=%lu, clockcache_page_size()=%lu\n",
```
Improved the message on L3058 to report something like:

```
3058       platform_error_log("%s():%d: Specified addr=%lu is not divisible by"
3059                          " configured page size, clockcache_page_size()=%lu\n",
```

Redefined this one fn as `static` but did not take on the task of fixing other such instances.

gapisback commented on src/clockcache.c:
```diff
@@ -3054,14 +3054,46 @@ clockcache_print(platform_log_handle *log_handle, clockcache *cc)
+   if (!retval) {
+      if (if_fails) {
+         platform_error_log(
+            "%s():%d: addr=%lu, base_addr=%lu, allocator_get_ref()=%d\n",
```
Expanded the erorr_log() with text describing the error condition.

gapisback commented on src/clockcache.c:
```diff
@@ -3054,14 +3054,46 @@ clockcache_print(platform_log_handle *log_handle, clockcache *cc)
+            base_addr,
+            (int)allocator_get_ref(cc->al, base_addr));
+      } else {
+         platform_error_log("%s():%d: addr=%lu, allocator_get_capacity()=%lu\n",
```
Done.

gapisback commented on src/platform_linux/laio.c:
```diff
@@ -231,7 +308,10 @@ laio_callback(io_context_t ctx, struct iocb *iocb, long res, long res2)
 }
 
-platform_status
+/*
+ * io_read_async() - Submit an Async read request.
+ */
+static platform_status
```
Added as requested.

gapisback commented on src/platform_linux/laio.h:
```diff
@@ -39,19 +46,27 @@ struct io_async_req {
    uint64           req_hand_base;
    uint64           req_hand[MAX_THREADS];
    platform_heap_id heap_id;
+   int              fd; // File descriptor to Splinter device
```
Changed as suggested.

rosenhouse commented on tests/functional/io_apis_test.c:
```diff
@@ -0,0 +1,832 @@
+                  master_cfg.io_flags,
+                  master_cfg.io_perms,
+                  master_cfg.io_async_queue_depth,
+                  "splinter_io_apis_test.db");
```
Could you choose a name that matches this glob?
https://github.com/vmware/splinterdb/blob/83de5e03722d61436ce247e406d87b8fa4913d85/.gitignore#L11

That way my git working tree stays clean when running this test.

gapisback replied:
Done.

rosenhouse replied:
@gapisback I'm sorry, but it still isn't fixed even with https://github.com/vmware/splinterdb/pull/473/commits/16738a6e21642d96c7b815f14b20f78c9140514f 

You'll need to have the filename start with `splinterdb` note the "db" at the end.

Alternately, change the `.gitignore` file to cover this file.

gapisback replied:
Oooops, sorry, @rosenhouse - I didn't quite get it the last time. Didn't read the `r.e` very carefully. Just pushed a commit to rename this to `splinterdb_io_apis_test_db`, which should sort out the problem.

rosenhouse replied:
Yup, fixed now, thanks!

gapisback commented:
@rosenhouse & @ajhconway - The last set of changes to address your feedback has been integrated. Please do clear this review before the end of your work day, so I can cleanup and check this into /main. Thanks!

gapisback commented:
HI, @ajhconway @rtjohnso -- I'd like to get your `LTGM` sign-off ... 

I have rebased v/s /main, picked up a new commit, squashed all my work to one commit and cleaned up the commit message.

If I don't hear from you by my Wed a.m. ... I'd like to push in to /main, as we've already discussed the changes in the PR meeting.

Thanks.
